![](download.png)

## Scenario
I am working on a data analysis project in Ethio clicks company. The CEO wants me to analayse a dataset which is related within the ethiopian context. The dataset that I chose for these project is a weather dataset that could be analysed and prepared to be trained for an AI based weather forecasting system.

## Step1: Importing important liberaries and load dataset


### Importing Liberay

In [2]:
#Import Important liberaries for these project
#! pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip
#import os
#os._exit(00)
import pandas as pd
# import pandas_profiling
# from pandas_profiling import ProfileReport
import numpy as np
import matplotlib.pyplot as plt
# import sklearn
# import pywedge as py

###  Loading Data

In [4]:
#Load Important DataSet For These Project
Weather_Data_Set = pd.read_csv('data/weatherHistory.csv')
#Display And View The DataSet in Tabular Form
Weather_Data_Set.head(2)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.

#### Inspect the top 5 data and the bottom 5 data

In [3]:
# The Top 5 data that are found in the weather dataset
Weather_Data_Set.head(5)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [4]:
# The Bottom 5 data that are found in the weather dataset
Weather_Data_Set.tail(5)

Formatted Date        Summary Precip Type  \
96448  2016-09-09 19:00:00.000 +0200  Partly Cloudy        rain   
96449  2016-09-09 20:00:00.000 +0200  Partly Cloudy        rain   
96450  2016-09-09 21:00:00.000 +0200  Partly Cloudy        rain   
96451  2016-09-09 22:00:00.000 +0200  Partly Cloudy        rain   
96452  2016-09-09 23:00:00.000 +0200  Partly Cloudy        rain   

       Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
96448        26.016667                 26.016667      0.43            10.9963   
96449        24.583333                 24.583333      0.48            10.0947   
96450        22.038889                 22.038889      0.56             8.9838   
96451        21.522222                 21.522222      0.60            10.5294   
96452        20.438889                 20.438889      0.61             5.8765   

       Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
96448                    31.0          16.1000         0.0   
96449                    20.0          15.5526         0.0   
96450                    30.0          16.1000         0.0   
96451                    20.0          16.1000         0.0   
96452                    39.0          15.5204         0.0   

       Pressure (millibars)                           Daily Summary  
96448               1014.36  Partly cloudy starting in the morning.  
96449               1015.16  Partly cloudy starting in the morning.  
96450               1015.66  Partly cloudy starting in the morning.  
96451               1015.95  Partly cloudy starting in the morning.  
96452               1016.16  Partly cloudy starting in the morning.

## Step2: Explore DataSet

###  Exploring Basic Information about the weather Dataset in traditional way.

In [5]:
# Basic Info About The Weather DataFrame
Weather_Data_Set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  object 
 1   Summary                   96453 non-null  object 
 2   Precip Type               95936 non-null  object 
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  object 
dtypes: float64(8), object(4)
memory usage: 8.8+ MB


In [6]:
# Basic Data Describe Method
Weather_Data_Set.describe()

Temperature (C)  Apparent Temperature (C)      Humidity  \
count     96453.000000              96453.000000  96453.000000   
mean         11.932678                 10.855029      0.734899   
std           9.551546                 10.696847      0.195473   
min         -21.822222                -27.716667      0.000000   
25%           4.688889                  2.311111      0.600000   
50%          12.000000                 12.000000      0.780000   
75%          18.838889                 18.838889      0.890000   
max          39.905556                 39.344444      1.000000   

       Wind Speed (km/h)  Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
count       96453.000000            96453.000000     96453.000000     96453.0   
mean           10.810640              187.509232        10.347325         0.0   
std             6.913571              107.383428         4.192123         0.0   
min             0.000000                0.000000         0.000000         0.0   
25%             5.828200              116.000000         8.339800         0.0   
50%             9.965900              180.000000        10.046400         0.0   
75%            14.135800              290.000000        14.812000         0.0   
max            63.852600              359.000000        16.100000         0.0   

       Pressure (millibars)  
count          96453.000000  
mean            1003.235956  
std              116.969906  
min                0.000000  
25%             1011.900000  
50%             1016.450000  
75%             1021.090000  
max             1046.380000

In [7]:
# Get a View of unique value in a column from a give dataset e.g Preciption Type
Weather_Data_Set['Precip Type'].unique()

array(['rain', 'snow', nan], dtype=object)

In [8]:
#Get a NULL value from each column
Weather_Data_Set.isnull().sum()

Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind Speed (km/h)             0
Wind Bearing (degrees)        0
Visibility (km)               0
Loud Cover                    0
Pressure (millibars)          0
Daily Summary                 0
dtype: int64

###### NB. From the above we can see that on precip type there are about 517 column which are null. We will fill those data with the mean value of that specific column.

In [9]:
Weather_Data_Set['Precip Type'].fillna(method="ffill",inplace=True)

In [10]:
Weather_Data_Set.isnull().sum()

Formatted Date              0
Summary                     0
Precip Type                 0
Temperature (C)             0
Apparent Temperature (C)    0
Humidity                    0
Wind Speed (km/h)           0
Wind Bearing (degrees)      0
Visibility (km)             0
Loud Cover                  0
Pressure (millibars)        0
Daily Summary               0
dtype: int64

#### NB. As you can see from above data, we can observe that we have successfully fill the data. i.e because as you see from above there is no record of the preciption column that has a null value, because the value of the precip type column is changed from 517 to 0.

### Exploring the dataset in Automated way.

In [11]:
Weather_Data_Set.profile_report()
profile =  ProfileReport(Weather_Data_Set,title='Birhan Data Analysis Project',html = { 'style' : {'full_width':True}})
profile.to_file(output_file='Birhan_Automated_Data-Analysis_Report.html')

Summarize dataset:   0%|          | 0/5 [00:02<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## Step3: Data Preparation and Analysis

### Tasks:
           .Provide the basic information about the dataframe.row,columns and types.
           .Finding out the different daily weather summary from our dataset.
           .Find the number of times when the Preciption type was exactly 'Rain'.
           .Find all the unique wind bearings(degrees).
           .Find the number of times when the wind speed was was exactly >5 km/h.
           .Display and view the humudity and pressure column from the given dataset.
           .Get the middle rows from the given dataset.
           .What are the first 10 humudity values that are recorded on Feb 07.
           .What are the top 10 apparent temperature value and their counts.
           .Find all the data values from the given dataset where visibility>15 and windspeed>5.
           .What is the mean apparent temprature value in a month using groupby.
           .Extract the pressure from the dataset and store it in a new column called 'New_Pressure_Column'.
           .Get the maximum temprature record that is recored when the precip type is 'snow'.

In [105]:
#Providing the basic information about the weather dataset
Weather_Data_Set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype              
---  ------                    --------------  -----              
 0   Formatted Date            96453 non-null  datetime64[ns, UTC]
 1   Summary                   96453 non-null  object             
 2   Precip Type               96453 non-null  object             
 3   Temperature (C)           96453 non-null  float64            
 4   Apparent Temperature (C)  96453 non-null  float64            
 5   Humidity                  96453 non-null  float64            
 6   Wind Speed (km/h)         96453 non-null  float64            
 7   Wind Bearing (degrees)    96453 non-null  float64            
 8   Visibility (km)           96453 non-null  float64            
 9   Loud Cover                96453 non-null  float64            
 10  Pressure (millibars)      96453 non-null  float64            
 11  Daily Summary  

In [13]:
#Find out the different daily weather condtion
Weather_Data_Set['Daily Summary'].unique()

array(['Partly cloudy throughout the day.',
       'Mostly cloudy throughout the day.', 'Foggy in the evening.',
       'Foggy overnight and breezy in the morning.',
       'Overcast throughout the day.', 'Partly cloudy until night.',
       'Mostly cloudy until night.',
       'Foggy starting overnight continuing until morning.',
       'Foggy in the morning.', 'Partly cloudy until evening.',
       'Partly cloudy starting in the morning.',
       'Mostly cloudy starting overnight continuing until night.',
       'Mostly cloudy until evening.',
       'Partly cloudy starting in the morning continuing until evening.',
       'Partly cloudy starting in the afternoon.',
       'Partly cloudy starting overnight.',
       'Partly cloudy until morning.',
       'Partly cloudy starting overnight continuing until night.',
       'Partly cloudy starting in the afternoon continuing until night.',
       'Mostly cloudy starting overnight.',
       'Partly cloudy until afternoon.',
       'Mostly

In [15]:
## Finding the no of time when the precip type was exactly rain
len(Weather_Data_Set[Weather_Data_Set['Precip Type']=='rain'])

85741

In [16]:
## Finding all the unique wind bearning value in the dataset
Weather_Data_Set['Wind Bearing (degrees)'].unique()

array([251., 259., 204., 269., 258., 260., 279., 290., 316., 281., 289.,
       262., 288., 230., 163., 139., 147., 160., 152., 150., 149., 180.,
       161., 135., 141., 151., 169., 170., 187., 179., 162., 159., 168.,
        32., 140., 103., 113., 129., 207., 153.,   4., 341.,  15., 348.,
       321., 311., 339., 340., 330.,  19., 277.,   9.,   0., 350., 349.,
       338., 320., 310., 328.,  20.,  28.,  11., 326., 309., 193., 273.,
       300., 307., 319., 318., 243., 177., 172., 142., 130., 359., 166.,
       145., 178., 223., 240., 231., 214., 222., 241., 235., 238., 211.,
       221., 215., 224., 358.,   8.,  59.,  63.,  65., 146., 305., 327.,
       271., 297., 301., 308., 272., 351., 175., 138., 158., 132., 209.,
       250., 295., 280., 270., 239., 242., 266., 278., 325., 282., 274.,
       255.,  46., 284., 283., 313., 345.,  16., 332.,  12.,  39.,   3.,
        33.,  24.,  25.,  31.,  47.,  67.,  60., 144.,  57.,   2.,  18.,
        48.,  29., 335., 228., 315.,  40., 143., 13

In [20]:
## Finding the no of wind speed values exactly >5 km/h
Weather_Data_Set[Weather_Data_Set['Wind Speed (km/h)']>5].value_counts().sum()

74553

In [33]:
# Display and view the pressure and humudity column from the given dataset
Pressure_Column = pd.DataFrame({'Pressure(millibars)':Weather_Data_Set['Pressure (millibars)']})
Humdidity_Column = pd.DataFrame({'Humidity':Weather_Data_Set['Humidity']})
Combined_Column = Pressure_Column.join(Humdidity_Column)
Combined_Column

Pressure(millibars)  Humidity
0                  1015.13      0.89
1                  1015.63      0.86
2                  1015.94      0.89
3                  1016.41      0.83
4                  1016.51      0.83
...                    ...       ...
96448              1014.36      0.43
96449              1015.16      0.48
96450              1015.66      0.56
96451              1015.95      0.60
96452              1016.16      0.61

[96453 rows x 2 columns]

In [34]:
# Get the value of the random middle values from a given dataset
Weather_Data_Set[3000:3010]

Formatted Date        Summary Precip Type  \
3000  2006-01-14 00:00:00.000 +0100          Foggy        snow   
3001  2006-01-14 01:00:00.000 +0100          Clear        snow   
3002  2006-01-14 02:00:00.000 +0100       Overcast        snow   
3003  2006-01-14 03:00:00.000 +0100       Overcast        snow   
3004  2006-01-14 04:00:00.000 +0100       Overcast        snow   
3005  2006-01-14 05:00:00.000 +0100       Overcast        snow   
3006  2006-01-14 06:00:00.000 +0100       Overcast        snow   
3007  2006-01-14 07:00:00.000 +0100          Foggy        snow   
3008  2006-01-14 08:00:00.000 +0100  Mostly Cloudy        snow   
3009  2006-01-14 09:00:00.000 +0100  Mostly Cloudy        snow   

      Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
3000        -0.116667                 -2.388889      0.92             6.6332   
3001        -3.605556                 -5.850000      0.92             5.3613   
3002        -0.650000                 -2.883333      0.96             6.3112   
3003        -0.650000                 -2.822222      0.92             6.1663   
3004        -0.605556                 -0.605556      0.93             3.2361   
3005        -0.650000                 -0.650000      0.96             3.3488   
3006        -0.650000                 -0.650000      0.92             3.2844   
3007        -0.172222                 -0.172222      0.87             4.2021   
3008        -0.144444                 -3.261111      0.86             9.2414   
3009        -0.166667                 -2.738889      0.86             7.4543   

      Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
3000                   318.0           2.9463         0.0   
3001                    35.0           3.6386         0.0   
3002                   320.0           3.3971         0.0   
3003                   349.0           3.4132         0.0   
3004                   344.0           3.3649         0.0   
3005                    14.0           3.3971         0.0   
3006                    22.0           3.3971         0.0   
3007                    56.0           2.9785         0.0   
3008                    58.0           4.3148         0.0   
3009                    80.0           4.7656         0.0   

      Pressure (millibars)                                      Daily Summary  
3000               1036.66  Mostly cloudy starting overnight continuing un...  
3001               1037.76  Mostly cloudy starting overnight continuing un...  
3002               1036.48  Mostly cloudy starting overnight continuing un...  
3003               1036.37  Mostly cloudy starting overnight continuing un...  
3004               1036.26  Mostly cloudy starting overnight continuing un...  
3005               1036.06  Mostly cloudy starting overnight continuing un...  
3006               1036.06  Mostly cloudy starting overnight continuing un...  
3007               1036.43  Mostly cloudy starting overnight continuing un...  
3008               1036.86  Mostly cloudy starting overnight continuing un...  
3009               1037.07  Mostly cloudy starting overnight continuing un...

In [87]:
#What is the humidity value that was recorded on july 07.
import datetime
Weather_Data_Set['Formatted Date'] = pd.to_datetime(Weather_Data_Set['Formatted Date'],utc=True)
Weather_Data_Set.loc[2006-1-14,'Humidity']

0.96

In [89]:
#What are the top 10 apparent temperature value and their counts.
Weather_Data_Set['Apparent Temperature (C)'].value_counts().sort_values(ascending=False).head(10)

12.777778    378
17.777778    373
13.888889    342
17.222222    330
12.222222    305
15.000000    301
16.111111    287
11.111111    279
22.777778    279
20.000000    271
Name: Apparent Temperature (C), dtype: int64

In [104]:
#Find all the data values from the given dataset where visibility>15 and windspeed>5.
Weather_data_Visibility=Weather_Data_Set[Weather_Data_Set['Visibility (km)']>15]
Weather_data_Windspeed = Weather_Data_Set[Weather_Data_Set['Wind Speed (km/h)']>5]

Visisbility_df = pd.DataFrame({"Visibility Data (>15)":Weather_data_Visibility['Visibility (km)']})
Windspeed_df = pd.DataFrame({"Windspeed Data (>5)":Weather_data_Windspeed['Wind Speed (km/h)']})

combined_df =Visisbility_df.join(Windspeed_df)
combined_df

Visibility Data (>15)  Windspeed Data (>5)
0                    15.8263              14.1197
1                    15.8263              14.2646
3                    15.8263              14.1036
4                    15.8263              11.0446
22                   15.8263               6.6493
...                      ...                  ...
96448                16.1000              10.9963
96449                15.5526              10.0947
96450                16.1000               8.9838
96451                16.1000              10.5294
96452                15.5204               5.8765

[20123 rows x 2 columns]

In [124]:
#Find the list of the apparent temprature value for Preciption type of 'Snow' using groupby.
Data2 = Weather_Data_Set.groupby("Precip Type").get_group('snow')
ApparentData = Data2['Apparent Temperature (C)']
SnowData = Data2['Precip Type']
Total_Data = pd.DataFrame({"Apparent Temp":ApparentData,
                          "Snow Data":SnowData})
Total_Data

Apparent Temp Snow Data
1562       -4.150000      snow
1563       -4.061111      snow
1564       -3.477778      snow
1565       -4.400000      snow
1566       -4.438889      snow
...              ...       ...
93265      -2.950000      snow
93266      -1.111111      snow
93267      -1.044444      snow
93311      -0.511111      snow
93506      -2.705556      snow

[10712 rows x 2 columns]

In [125]:
# Finding the mean value for the above data
Total_Data['Apparent Temp'].mean()

-6.2498470043979735

In [130]:
# Extract the pressure from the dataset and store it in a new column called 'New_Pressure_Column'
Weather_Data_Set['New_Pressure_Column'] = Weather_Data_Set['Pressure (millibars)']*100
Weather_Data_Set.head(2)

Formatted Date        Summary Precip Type  Temperature (C)  \
0 2006-03-31 00:00:00+00:00  Partly Cloudy        rain         9.472222   
1 2006-03-31 00:00:00+00:00  Partly Cloudy        rain         9.355556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   

                       Daily Summary  New_Pressure_Column  
0  Partly cloudy throughout the day.             101513.0  
1  Partly cloudy throughout the day.             101563.0

In [140]:
#Get the maximum temprature record that is recored when the precip type is 'snow'.
Data3 = Weather_Data_Set.groupby("Precip Type").get_group('snow')
Data3['Temperature (C)'].max()

0.0

## Step4: Further analyis and Deep Dive Data Visualaiztion

#### Basic tasks to be performed:
              .We will try to analyse the data using the visualiztion graphs.
              .We will try to visualize the data using charts,histograms and other optional visualization graphs.

In [15]:
#Dataset Visualiztion
visualyser = py.Pywedge_Charts(Weather_Data_Set,c=None, y='Summary')

C:\Users\GL\Desktop\sample_project_1\env\lib\site-packages\pywedge\pywedge.py:27: FutureWarning:

In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.



In [16]:
Dataset_Visualiztion_Graphs=visualyser.make_charts()

HTML(value='<h2>Pywedge Make_Charts </h2>')